include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+        else if (from == dim && to == dim - 1)
+          return cell_entities;
+        else if (dim == 3 && from == 2 && to == 0)
+          return quad_vertices;
+        else if (dim == 3 && from == 2 && to == 1)
+          return quad_lines;
+        else if (from == 1 && to == 0)
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+      inline const CRS<T> &
+      entity_to_entities(const unsigned int from, const unsigned int to) const
+      {
+        if (from == dim && to == dim)
+          return neighbors;
+        else if (from == dim && to == dim - 1)
+          return cell_entities;
+        else if (dim == 3 && from == 2 && to == 0)
+          return quad_vertices;
+        else if (dim == 3 && from == 2 && to == 1)
+          return quad_lines;
+        else if (from == 1 && to == 0)
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+    private:
+      const unsigned int        dim;
+      std::vector<unsigned int> cell_types;
+
+      CRS<T> line_vertices;
+
+      std::vector<unsigned char> line_orientation;
+
+      CRS<T> quad_vertices;
+      CRS<T> quad_lines;
+
+      std::vector<unsigned char> quad_orientation;
+
+      CRS<T> cell_entities;
+      CRS<T> neighbors;
+
+      std::vector<unsigned int> quad_types;
+    };
+
+
+
+    /**
+     * Create the transposed of a compressed row storage matrix.
+     */
+    template <typename T>
+    CRS<T>
+    transpose(const CRS<T> &con)
+    {
+      auto        con_copy = con;
+      auto &      col      = con_copy.col;
+      const auto &ptr      = con_copy.ptr;
+
+      CRS<T> con_t;
+      auto & col_t = con_t.col;
+      auto & ptr_t = con_t.ptr;
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        std::sort(col.data() + ptr[i], col.data() + ptr[i + 1]);
+
+      col_t.clear();
+      ptr_t.clear();
+
+      col_t.reserve(col.size());
+      ptr_t.reserve(
+        col.size() == 0 ? 0 : *std::max_element(col.begin(), col.end()));
+
+      std::vector<std::pair<T, T>> temp;
+      temp.reserve(col.size());
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        for (std::size_t j = ptr[i]; j < ptr[i + 1]; j++)
+          temp.emplace_back(col[j], i);
+
+      std::sort(temp.begin(), temp.end());
+
+      T r = -1;
+
+      for (unsigned int i = 0; i < temp.size(); i++)
+        {
+          if (r != temp[i].first)
+            {
+              r = temp[i].first;
+              ptr_t.push_back(col_t.size());
+            }
+
+          col_t.push_back(temp[i].second);
+        }
+
+      ptr_t.push_back(col_t.size());
+
+      return con_t;
+    }
+
+
+
+    /**
+     * Determine the neighbors of a cell by visiting its faces and looking for
+     * a cell which is not equal to this cell.
+     */
+    template <typename T>
+    CRS<T>
+    determine_neighbors(const CRS<T> &con_cf, const CRS<T> &con_fc)
```

Comment:
Even though this is an internal function where we do not set as high a bar on documentation, I would like some more expressive variables name to see what should happen here.

Reply:
Done!